File: exercise2/VB/distribution_normal/Form1.vb
Imports System
Imports System.Timers
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports System.Threading
Imports System.Security.Cryptography
Imports System.Collections
Imports System.Diagnostics
Imports System.Windows.Forms.VisualStyles.VisualStyleElement.TaskbarClock

Namespace session2
    Partial Public Class Form1
        Inherits Form

        Public random As Random = New System.Random()
        Private counts As Dictionary(Of Double, Integer) = New Dictionary(Of Double, Integer)()

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
        End Sub

        Private Sub button2_Click(ByVal sender As Object, ByVal e As EventArgs)
            timer1.[Stop]()
            Dim sum As Integer = 0
            Dim prob As Dictionary(Of Double, Double) = New Dictionary(Of Double, Double)()

            For Each kvp As KeyValuePair(Of Double, Integer) In counts
                sum += kvp.Value
            Next

            For Each kvp As KeyValuePair(Of Double, Integer) In counts
                Dim probability As Double = CDbl(kvp.Value) / sum
                Dim x = Math.Round(probability, 2)
                prob.Add(kvp.Key, x)
            Next

            Dim output As String = ""

            For Each kvp As KeyValuePair(Of Double, Double) In prob
                output += String.Format("Number = {0}, probability = {1}", kvp.Key, kvp.Value)
                output += vbLf
            Next

            MessageBox.Show(output.ToString())
        End Sub

        Private Sub timer1_Tick(ByVal sender As Object, ByVal e As EventArgs)
            Dim i = 1
            Dim random = New Random()
            Dim rDouble = random.NextDouble()
            Dim r = random.NextDouble()
            Dim d = Math.Round(r, 2)
            Dim keyExists As Boolean = counts.ContainsKey(d)

            If keyExists Then
                counts(d) += 1
            Else
                counts.Add(d, i)
            End If

            richTextBox1.Text += Environment.NewLine & d
        End Sub

        Private Sub richTextBox1_TextChanged(ByVal sender As Object, ByVal e As EventArgs)
        End Sub

        Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs)
            timer1.Interval = 2000
            timer1.Start()
        End Sub

        Public Shared Function SampleGaussian(ByVal random As Random, ByVal mean As Double, ByVal stddev As Double) As Double
            Dim x1 As Double = 1 - random.NextDouble()
            Dim x2 As Double = 1 - random.NextDouble()
            Dim y1 As Double = Math.Sqrt(-2.0 * Math.Log(x1)) * Math.Cos(2.0 * Math.PI * x2)
            Return y1 * stddev + mean
        End Function

        Friend WithEvents Timer1 As Windows.Forms.Timer
        Private components As IContainer

        Private Sub InitializeComponent()
            Me.components = New System.ComponentModel.Container()
            Me.Timer1 = New System.Windows.Forms.Timer(Me.components)
            Me.RichTextBox1 = New System.Windows.Forms.RichTextBox()
            Me.SuspendLayout()
            '
            'RichTextBox1
            '
            Me.RichTextBox1.Location = New System.Drawing.Point(157, 127)
            Me.RichTextBox1.Name = "RichTextBox1"
            Me.RichTextBox1.Size = New System.Drawing.Size(314, 96)
            Me.RichTextBox1.TabIndex = 0
            Me.RichTextBox1.Text = ""
            '
            'Form1
            '
            Me.AutoScaleDimensions = New System.Drawing.SizeF(7.0!, 15.0!)
            Me.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Font
            Me.ClientSize = New System.Drawing.Size(800, 450)
            Me.Controls.Add(Me.RichTextBox1)
            Me.Name = "Form1"
            Me.Text = "Form1"
            Me.ResumeLayout(False)

        End Sub

        Friend WithEvents RichTextBox1 As RichTextBox
    End Class
End Namespace

File: exercise2/c#/session2/Form1.cs
using System;
using System.Timers;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Threading;
using System.Security.Cryptography;
using System.Collections;
using System.Diagnostics;

namespace session2
{
    public partial class Form1 : Form
    {
        public Random random = new System.Random();
        Dictionary<double, int> counts = new Dictionary<double, int>();
        public Form1()
        {
            InitializeComponent();
            

        }

        private void Form1_Load(object sender, EventArgs e)
        {
          
        }

        private void button2_Click(object sender, EventArgs e)
        {
            timer1.Stop();
            int sum = 0;
            Dictionary<double, double> prob = new Dictionary<double, double>();
            //make sum of all values in dictionary
            //probability for each double number
            foreach (KeyValuePair<double, int> kvp in counts)
            {
                sum += kvp.Value;
                
            }
            
            foreach (KeyValuePair<double, int> kvp in counts)
            {
                double probability = (double)kvp.Value / sum;
                var x = Math.Round(probability, 2);
                prob.Add(kvp.Key, x);
            }
            

            // print the dictionary to console
            string output = "";
            foreach (KeyValuePair<double, double> kvp in prob)
            {
               output += string.Format("Number = {0}, probability = {1}", kvp.Key, kvp.Value);
               output += "\n";
            }
            MessageBox.Show(output.ToString());





        }

        private void timer1_Tick(object sender, EventArgs e)
        {
            

            var i = 1;
            
            var random = new Random();
            var rDouble = random.NextDouble();
           
            var r = random.NextDouble();
            var d = Math.Round(r, 2);

            bool keyExists = counts.ContainsKey(d);
            if (keyExists)
            {
                counts[d] += 1;

            }
            else
            {
                counts.Add(d, i);
            }

            //counts.Add(d, 0);
            richTextBox1.Text += Environment.NewLine + d;

           


        }

        private void richTextBox1_TextChanged(object sender, EventArgs e)
        {

        }

        private void button1_Click(object sender, EventArgs e)
        {

           
            timer1.Interval = 2000;
            timer1.Start();
           
        }
        public static double SampleGaussian(Random random, double mean, double stddev)
        {
            double x1 = 1 - random.NextDouble();
            double x2 = 1 - random.NextDouble();

            double y1 = Math.Sqrt(-2.0 * Math.Log(x1)) * Math.Cos(2.0 * Math.PI * x2);
            return y1 * stddev + mean;
        }

    }
}
